### PR TITLE
GH-1285: [COVID] Launch submissions through Rawls, and keep track of them

### PR DESCRIPTION
**Pull request by tbl3rd**
### Purpose

- https://broadinstitute.atlassian.net/browse/GH-1285

### Changes

- Implement start-covid-workload! and add a unit test that mocks create-covid-workload!.
- Clean up some time interval arithmetic and add a snapshots test for it.

### Review Instructions

- Start with covid_test.clj, then covid.clj, then look at snapshots_test.clj and its subjects.
- Tell me what to throw away if you're sure we won't need it.

### CODE REVIEW

**Review by rexwangcc — APPROVED** [review]

File: api/src/wfl/service/rawls.clj [hunk 1]
```diff
@@ -88,3 +112,50 @@
                     :body         (json/write-str (map make-request entities)
                                                   :escape-slash false)})
         util/response-body-json)))
+
+;; https://cromwell.readthedocs.io/en/stable/execution/ExecutionTwists/
+;; :NoNewCalls is the default.
+;;
+(def a-rawls-submission
```
Comment (okotsopoulos) [inline api/src/wfl/service/rawls.clj:119]: Is this used anywhere?  What is its purpose?

File: api/src/wfl/module/covid.clj [hunk 2]
```diff
@@ -14,19 +14,32 @@
...
 
+(defn ^:private get-snapshots-from-workspace
+  [workspace])
+
 (defn start-covid-workload!
-  [])
+  "Mark WORKLOAD with a started timestamp."
+  [tx {:keys [id started] :as workload}]
+  (jdbc/update! tx :workload {:started (OffsetDateTime/now)} ["id = ?" id])
+  (workloads/load-workload-for-id tx id))
```
Comment (ehigham) [inline api/src/wfl/module/covid.clj:42]: Minor optimisation here, do we need to re-load the workload or can we just update the `:started` field?

File: api/src/wfl/service/rawls.clj [hunk 3]
```diff
@@ -88,3 +112,50 @@
...
+(defn submission-status
+  "Status for SUBMISSION-ID in WORKSPACE name/namespace."
+  [workspace submission-id]
+  {:pre (spec/uuid-string? submission-id)}
+  (-> {:method       :get
+       :url          (workspace-api-url workspace "submissions" submission-id)
+       :headers      (auth/get-auth-header)
+       :content-type :application/json}
+      http/request
+      util/response-body-json))
```
Comment (ehigham) [inline api/src/wfl/service/rawls.clj:161]: Can you put these in `firecloud`? Some of this functionality already exists there.

File: api/test/wfl/tools/snapshots.clj [hunk 4]
```diff
@@ -4,29 +4,23 @@
...
-  (let [columns     (-> (datarepo/all-columns dataset table)
-                        (->> (map :name) set)
-                        (conj "datarepo_row_id"))]
+  (let [columns (-> (datarepo/all-columns dataset (name table))
+                    (->> (map :name) set)
+                    (conj "datarepo_row_id"))]
     (-> (datarepo/make-snapshot-request dataset columns table row-ids)
         (update :name util/randomize)
-        (update :profileId (constantly tdr-profile)))))
+        (assoc :profileId tdr-profile))))
```
Comment (ehigham) [inline api/test/wfl/tools/snapshots.clj:14]: simpler. nice.

File: api/test/wfl/unit/snapshots_test.clj [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+                                      dataset table :updated
+                                      (util/days 40)
+                                      [:datarepo_row_id])
+                                     :rows flatten
+                                     (->> (snapshots/unique-snapshot-request
+                                           profile dataset table))))]
+        (let [days40     (snap (util/days 40))
+              march      (snap ["2021-03-29 00:00:00" "2021-03-31 00:00:00"])
+              [_ _ both] (data/diff days40 march)]
+          (is (apply = (cons both (map unname [days40 march])))))))))
```
Comment (ehigham) [inline api/test/wfl/unit/snapshots_test.clj:25]: this code already exists in the datarepo integration test Please don't talk to external services in a unit test.

File: api/test/wfl/unit/modules/covid_test.clj [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            [clojure.string      :as str]
+            [wfl.jdbc            :as jdbc]
+            [wfl.module.covid    :as covid]
+            [wfl.tools.fixtures  :as fixtures]
+            [wfl.tools.workloads :as workloads]
+            [wfl.wfl             :as wfl])
+  (:import [java.time OffsetDateTime]
+           [java.util UUID]))
+
+(use-fixtures :once fixtures/temporary-postgresql-database)
```
Comment (ehigham) [inline api/test/wfl/unit/modules/covid_test.clj:13]: likewise database usage generally means an integration test.

File: api/src/wfl/service/datarepo.clj [hunk 7]
```diff
@@ -225,21 +225,24 @@
...
 (defn ^:private query-table-between-impl
-  [{:keys [dataProject] :as dataset} table column-name [start end] col-spec]
-  (let [bq-name (bigquery-name dataset)
-        query   "SELECT %s
-                 FROM   `%s.%s.%s`
-                 WHERE %s BETWEEN '%s' AND '%s'"]
-    (->> (format query col-spec dataProject bq-name table column-name start end)
-         (bigquery/query-sync dataProject))))
+  [{:keys [dataProject] :as dataset} table between [start end] col-spec]
+  (let [[table between] (map name [table between])
```
[the remark this answers is not in the document]
Comment (tbl3rd) [reply api/src/wfl/service/datarepo.clj:229]: Some things can be keywords and some strings.
We can force them all one way or the other, but
Clojure JDBC likes table names to be keywords,
and so on, so just treat them all consistently?

File: api/src/wfl/service/rawls.clj [hunk 8]
```diff
@@ -44,8 +45,31 @@
...
   (get-workspace-json workspace "snapshots" reference-id))
 
-(defn delete-snapshot-reference
-  "Delete the snapshot reference in fully-qualified Terra WORKSPACE with REFERENCE-ID."
+(defn snapshots
+  "Return a lazy sequence of snapshots in WORKSPACE namespace/name."
+  [workspace]
+  (let [rawls   "https://rawls.dsde-prod.broadinstitute.org/api/workspaces"
+        url     (str/join "/" [rawls workspace "snapshots"])
+        limit   23
```
[the remark this answers is not in the document]
Comment (tbl3rd) [reply api/src/wfl/service/rawls.clj:53]: Why not?  We'll probably delete this def anyway.

File: api/src/wfl/service/rawls.clj (hunk 1 above)
Comment (tbl3rd) [reply api/src/wfl/service/rawls.clj:119]: To remind me what a Rawls submission looks like.

File: api/src/wfl/module/covid.clj (hunk 2 above)
Comment (tbl3rd) [reply api/src/wfl/module/covid.clj:42]: `start-` things tend to return the workload, but we don't have to, of course.

File: api/src/wfl/service/rawls.clj (hunk 3 above)
Comment (tbl3rd) [reply api/src/wfl/service/rawls.clj:161]: Yeah.  I think those came from the mob branch.
Maybe get rid of them.

File: api/test/wfl/unit/modules/covid_test.clj (hunk 6 above)
Comment (tbl3rd) [reply api/test/wfl/unit/modules/covid_test.clj:13]: OK.  I could probably mock more, but then we aren't really testing anything.

File: api/src/wfl/module/covid.clj [hunk 9]
```diff
@@ -14,19 +14,32 @@
             [wfl.service.rawls :as rawls]
             [wfl.util :as util]
             [wfl.wfl :as wfl])
-  (:import (java.time OffsetDateTime)))
+  (:import [java.time OffsetDateTime]))
 
 (def pipeline "Sarscov2IlluminaFull")
 
+(defn row-ids-yesterday
```
[the remark this answers is not in the document]
Comment (tbl3rd) [reply api/src/wfl/module/covid.clj:21]: OK.

**Pull request merged**